express-notes-api/server.js

```diff
@@ -0,0 +1,111 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+
+const app = express();
+
+app.use(express.json());
+
+// reusable functions
+async function read() {
+  const data = await readFile('data.json', 'utf8');
+  return JSON.parse(data);
+}
+
+async function write(data, res) {
+  await writeFile('data.json', JSON.stringify(data, null, 2));
+}
+
+function handleError(res, err) {
+  console.error(err);
+  res.status(500).json({ error: 'An unexpected error occurred.' });
+}
+//
+
+app.get('/api/notes', async (req, res) => {
+  try {
+    const data = await read();
+    const notes = [];
+    for (const entry in data.notes) {
+      notes.push(data.notes[entry]);
+    }
+    res.status(200).json(notes);
+  } catch (err) {
+    handleError(res, err);
+  }
+});
+
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const data = await read();
+    const id = Number(req.params.id);
+    if (id < 1 || isNaN(id) || !Number.isInteger(id)) {
+      res.status(400).json({ error: 'id must be a positive integer' });
+    } else if (!data.notes[id]) {
+      res.status(404).json({ error: `cannot find note with id ${id}` });
+    } else {
+      res.status(200).json(data.notes[id]);
+    }
+  } catch (err) {
+    handleError(res, err);
+  }
+});
+
+app.post('/api/notes', async (req, res) => {
+  const bodyContent = req.body.content;
+  try {
+    if (!bodyContent) {
+      res.status(400).json({ error: `content is a required field` });
+    }
```

Comment:
You need to make sure you return without executing any other code after your return an error response. Sending an error response does not terminate execution of the function. So this will send an error and then attempt to do the read, write, and sending the success response. It should result in an exception from Express that you are modifying the response after the headers have been set.

You'll need to make this fix in all places where you return an error.